### ISSUE TITLE
unprivileged container with xattr support for samba AD

### ISSUE THREAD

**Issue by bekkibobbek** [lxc/lxc#2708]
The template below is mostly useful for bug reports and support questions.
Feel free to remove anything which doesn't apply to you and add more information where it makes sense.

# Required information

 * Distribution: Ubuntu
 * Distribution version: 16.04.5 LTS
 * The output of
   * `lxc-start --version`: 2.0.8
   * `lxc-checkconfig`
```
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled
Multiple /dev/pts instances: enabled

--- Control groups ---
Cgroup: enabled
Cgroup clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled
Macvlan: enabled
Vlan: enabled
Bridges: enabled
Advanced netfilter: enabled
CONFIG_NF_NAT_IPV4: enabled
CONFIG_NF_NAT_IPV6: enabled
CONFIG_IP_NF_TARGET_MASQUERADE: enabled
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled
FUSE (for use with lxcfs): enabled

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: enabled
```
   * `uname -a`: 
```
Linux bob 4.4.0-134-generic #160-Ubuntu SMP Wed Aug 15 14:58:00 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

   * `cat /proc/self/cgroup`
```
11:blkio:/
10:freezer:/
9:hugetlb:/
8:perf_event:/
7:devices:/
6:memory:/
5:pids:/
4:net_cls,net_prio:/
3:cpu,cpuacct:/
2:cpuset:/
1:name=systemd:/
```
   * `cat /proc/1/mounts`
```
/dev/sda1 / btrfs rw,relatime,space_cache,user_subvol_rm_allowed,subvolid=833,subvol=/lxc/bob/rootfs 0 0
none /dev tmpfs rw,nodev,relatime,size=492k,mode=755,uid=427680,gid=427680 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
proc /proc/sys/net proc rw,nosuid,nodev,noexec,relatime 0 0
proc /proc/sys proc ro,nosuid,nodev,noexec,relatime 0 0
proc /proc/sysrq-trigger proc ro,nosuid,nodev,noexec,relatime 0 0
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
sysfs /sys sysfs ro,nosuid,nodev,noexec,relatime 0 0
sysfs /sys/devices/virtual/net sysfs rw,nodev,relatime 0 0
sysfs /sys/devices/virtual/net sysfs rw,nosuid,nodev,noexec,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
mqueue /dev/mqueue mqueue rw,nodev,relatime 0 0
udev /dev/full devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
udev /dev/null devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
udev /dev/random devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
udev /dev/tty devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
udev /dev/urandom devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
udev /dev/zero devtmpfs rw,nosuid,relatime,size=6130688k,nr_inodes=1532672,mode=755 0 0
binfmt_misc /proc/sys/fs/binfmt_misc binfmt_misc rw,relatime 0 0
lxcfs /proc/cpuinfo fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/diskstats fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/meminfo fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/stat fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/swaps fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lxcfs /proc/uptime fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
devpts /dev/console devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
devpts /dev/pts devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
devpts /dev/ptmx devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
devpts /dev/tty1 devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
devpts /dev/tty2 devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
devpts /dev/tty3 devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
devpts /dev/tty4 devpts rw,relatime,gid=427685,mode=620,ptmxmode=666 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev,uid=427680,gid=427680 0 0
tmpfs /run tmpfs rw,nosuid,nodev,mode=755,uid=427680,gid=427680 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k,uid=427680,gid=427680 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755,uid=427680,gid=427680 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
```

# Issue description
The issue is already described here:
https://lists.linuxcontainers.org/pipermail/lxc-devel/2015-November/012789.html
TL;DR
```
samba-tool domain provision --use-rfc2307 --server-role=dc --dns-backend=SAMBA_INTERNAL --realm=BOBB.BOB --domain=BOB --adminpass=$P 
You are not root or your system do not support xattr, using tdb backend for attributes. 
not using extended attributes to store ACLs and other metadata. If you intend to use this provision in production, rerun the script as root on a system supporting xattrs.
Looking up IPv4 addresses
Looking up IPv6 addresses
No IPv6 address will be assigned
Setting up secrets.ldb
Setting up the registry
Setting up the privileges database
Setting up idmap db
Setting up SAM db
Setting up sam.ldb partitions and settings
Setting up sam.ldb rootDSE
Pre-loading the Samba 4 and AD schema
Adding DomainDN: DC=bobb,DC=bob
Adding configuration container
Setting up sam.ldb schema
Setting up sam.ldb configuration data
Setting up display specifiers
Modifying display specifiers
Adding users container
Modifying users container
Adding computers container
Modifying computers container
Setting up sam.ldb data
Setting up well known security principals
Setting up sam.ldb users and groups
Setting up self join
process_usershare_file: share name unknown service (snum == -1) contains invalid characters (any of %<>*?|/\+=;:",)
xattr_tdb_removexattr() failed to get vfs_handle->data!
process_usershare_file: share name unknown service (snum == -1) contains invalid characters (any of %<>*?|/\+=;:",)
Security context active token stack underflow!
PANIC (pid 29901): Security context active token stack underflow!
BACKTRACE: 43 stack frames:
 #0 /usr/lib/x86_64-linux-gnu/samba/libsmbregistry.so.0(log_stack_trace+0x1a) [0x7fb37cb3b7aa]
 #1 /usr/lib/x86_64-linux-gnu/samba/libsmbregistry.so.0(smb_panic_s3+0x20) [0x7fb37cb3b880]
 #2 /usr/lib/x86_64-linux-gnu/libsamba-util.so.0(smb_panic+0x2f) [0x7fb38da67f1f]
 #3 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(sec_ctx_active_token+0x6a) [0x7fb3798fbf3a]
 #4 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(try_chown+0xa9) [0x7fb3799079f9]
 #5 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(set_nt_acl+0x1fd) [0x7fb379907c4d]
 #6 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(+0x1cf051) [0x7fb3799ca051]
 #7 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(smb_vfs_call_fset_nt_acl+0x2d) [0x7fb37990025d]
 #8 /usr/lib/x86_64-linux-gnu/samba/vfs/acl_xattr.so(+0x2389) [0x7fb366981389]
 #9 /usr/lib/x86_64-linux-gnu/samba/vfs/acl_xattr.so(+0x4416) [0x7fb366983416]
 #10 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(smb_vfs_call_fset_nt_acl+0x2d) [0x7fb37990025d]
 #11 /usr/lib/python2.7/dist-packages/samba/samba3/smbd.so(+0x2334) [0x7fb379d27334]
 #12 /usr/bin/python2.7(PyEval_EvalFrameEx+0x79bd) [0x4c37ed]
 #13 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #14 /usr/bin/python2.7(PyEval_EvalFrameEx+0x58b7) [0x4c16e7]
 #15 /usr/bin/python2.7(PyEval_EvalFrameEx+0x553f) [0x4c136f]
 #16 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #17 /usr/bin/python2.7(PyEval_EvalFrameEx+0x58b7) [0x4c16e7]
 #18 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #19 /usr/bin/python2.7(PyEval_EvalFrameEx+0x58b7) [0x4c16e7]
 #20 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #21 /usr/bin/python2.7() [0x4d55f3]
 #22 /usr/bin/python2.7(PyObject_Call+0x3e) [0x4a577e]
 #23 /usr/bin/python2.7(PyEval_EvalFrameEx+0x2f0d) [0x4bed3d]
 #24 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #25 /usr/bin/python2.7() [0x4d54b9]
 #26 /usr/bin/python2.7(PyObject_Call+0x3e) [0x4a577e]
 #27 /usr/bin/python2.7(PyEval_EvalFrameEx+0x2f0d) [0x4bed3d]
 #28 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #29 /usr/bin/python2.7() [0x4d54b9]
 #30 /usr/bin/python2.7(PyObject_Call+0x3e) [0x4a577e]
 #31 /usr/bin/python2.7(PyEval_EvalFrameEx+0x2f0d) [0x4bed3d]
 #32 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #33 /usr/bin/python2.7() [0x4d54b9]
 #34 /usr/bin/python2.7(PyObject_Call+0x3e) [0x4a577e]
 #35 /usr/bin/python2.7(PyEval_EvalFrameEx+0x2f0d) [0x4bed3d]
 #36 /usr/bin/python2.7(PyEval_EvalCodeEx+0x306) [0x4b9ab6]
 #37 /usr/bin/python2.7() [0x4eb30f]
 #38 /usr/bin/python2.7(PyRun_FileExFlags+0x82) [0x4e5422]
 #39 /usr/bin/python2.7(PyRun_SimpleFileExFlags+0x186) [0x4e3cd6]
 #40 /usr/bin/python2.7(Py_Main+0x612) [0x493ae2]
 #41 /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7fb38ed05830]
 #42 /usr/bin/python2.7(_start+0x29) [0x4933e9]
Can not dump core: corepath not set up
```

# Steps to reproduce

 1. Run unprivileged container
 2. Run `samba-tool domain provision --use-rfc2307 --interactive`

# Information to attach

 - [ ] any relevant kernel output (`dmesg`)
 - [ ] container log (The <log> file from running `lxc-start -n <c> -l <log> -o DEBUG`)
 - [ ] the containers configuration file

**Comment by brauner** [comment]
As noted in the link you posted here this seems like a bug in sambas tdb implementation. It seems it gets confused when it is run in unprivileged containers. There's nothing for LXC to do I fear. Sorry for the delayed reply.

**Comment by bradsoto** [comment]
I was able to support both unpriv containers and xattr with a few tweaks to the container and Samba's provisioning. Samba expects to be able to use UID 3000000+ (we only get 65534 in unpriv), and containers backed by a block device don't mount root with xattr support (unless they are btrfs).

### Making Samba Use UID 655+ instead of 3000000+
Edit the file /usr/share/samba/setup/idmap_init.ldif
```
dn: CN=CONFIG
cn: CONFIG
lowerBound: 655
upperBound: 65533

dn: @INDEXLIST
@IDXATTR: xidNumber
@IDXATTR: objectSid

dn: CN=S-1-5-32-544
cn: S-1-5-32-544
objectClass: sidMap
objectSid: S-1-5-32-544
type: ID_TYPE_BOTH
xidNumber: 655
distinguishedName: CN=S-1-5-32-544
```
Now provision with something like
```
samba-tool domain provision --realm=$(dnsdomainname) --domain=$(hostname) \
--host-name=$(hostname).$(dnsdomainname) --option="vfs objects = acl_xattr xattr_tdb" \
--use-xattrs=no --option="idmap config * : range = 655-65533" \
--option="dns forwarder = 1.1.1.1"
```
This will get you started. I haven't tested it much in production environments. And I am unsure if we need all the idmap config. Maybe someone from the Samba Team can cut in here. We now have an unpriv Samba AD container! BTW The Samba Team recommends using xattr_tbd. But if you _really_ want it:
### Add xattr support to LXC
Find where your image file is stored. It might be something like
```
/dev/rbd/{pool}/vm-{id}-disk-0
```
It's  most likely formatted ext4, but doesn't give us xattr even after running tune2fs -o ^user_xattr. Just convert it to btrfs with btrfs-convert. Then provision:
```
samba-tool domain provision --realm=$(dnsdomainname) --domain=$(hostname) \
--host-name=$(hostname).$(dnsdomainname) --option="vfs objects = btrfs acl_xattr" \
 --use-xattrs=yes --option="idmap config * : range = 655-65533" \
--option="dns forwarder = 1.1.1.1"
```

**Comment by paulolieuthier** [comment]
@bradsoto I can't reproduce your results. I guess the `--use-xattrs` option is no longer present. Are you still able to run samba in an unprivileged container with a recent samba version?

**Comment by albert-a** [comment]
A guy managed to provision AD-DC in an unprivileged LXC with a list of command line options:
https://lists.samba.org/archive/samba/2021-February/234326.html
although with some mapping related messages: `map_smb4_to_nfs4_id: Unknown gid [...]`

**Comment by cryptomilk** [comment]
## How to fix the issue?

The problem is the `acl_xattr:security_acl_name` default config option which doesn't work in an unprivileged container (`man vfs_acl_xattr`).

You need to use a different location and be careful with permissions, read the manpage especially the security warnings.

**Comment by SphtKr** [comment]
Scratch that--@cryptomilk's solution works great in fact. (My problem was actually not LXC related, I was trying to do the same thing in microk8s and had turned on the Kata runtime for the container at some point--which was apparently interfering with _any_ xattr usage in bind mounts--sorry for the off-topic post, too many issues open in tabs!)

**Comment by albert-a** [comment]
Hi, can somebody share a working solution?

@cryptomilk, thanks for the suggestion, but unfortunately it does not work for me. It seems that `acl_xattr:security_acl_name = user.NTACL` option just does not do anything, I always get the same error as OP.
Although I can verify that setting user.NTACL attribute on an arbitrary file is working in the container as shown in https://wiki.samba.org/index.php/File_System_Support:
```
touch test.txt
setfattr -n user.test -v test test.txt
getfattr -d test.txt
```
I see that the attribute `user.NTACL` is set and displayed.
- I tried to use `posix:eadb = /var/lib/samba/private/eadb.tdb`
- I tride setting `lxc.apparmor.profile: unconfined`
- I tried builtin bookworm version: 4.17.12 as well as version from the testing repository: 4.20.1
- I tried EXT4 and ZFS

None of these changed a thing, the same error.

I use unprivileged LXC container with nesting on Proxmox.
```
pve-manager/8.2.2/9355359cd7afbae4 (running kernel: 6.8.4-3-pve)
lxc-pve: 6.0.0-1
lxcfs: 6.0.0-pve2
zfsutils-linux: 2.2.3-pve2
```
Provision command:
```
samba-tool domain provision \
    --option="dns forwarder=$dns_forwarder" \
    --realm="$realm" --domain="$domain" --server-role=dc \
    --use-rfc2307 --host-name="$HOSTNAME" --host-ip=$ip --adminpass="$adminpass" \
    --option="acl_xattr:security_acl_name = user.NTACL"
```
Last log lines with keyword `acl` from `strace`, maybe it's related:
```
newfstatat(26, "", {st_mode=S_IFDIR|0755, st_size=4096, ...}, AT_EMPTY_PATH) = 0
fgetxattr(26, "user.NTACL", 0x4a5b060, 4096) = -1 ENODATA (No data available)
newfstatat(26, "", {st_mode=S_IFDIR|0755, st_size=4096, ...}, AT_EMPTY_PATH) = 0
fgetxattr(26, "system.posix_acl_access", 0x7fff999772a0, 132) = -1 ENODATA (No data available)
newfstatat(26, "", {st_mode=S_IFDIR|0755, st_size=4096, ...}, AT_EMPTY_PATH) = 0
getxattr("/proc/self/fd/26", "system.posix_acl_default", 0x7fff999772a0, 132) = -1 ENODATA (No data available)
newfstatat(AT_FDCWD, "/proc/self/fd/26", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
umask(000)                              = 022
```
